src/loki/reach.cc

```diff
@@ -110,17 +131,117 @@ directed_reach SimpleReach(const DirectedEdge* edge,
         continue;
       const auto* node = tile->node(edge.endnode());
       const auto* opp_edge = tile->directededge(node->edge_index() + edge.opp_index());
+
+      // NOTE: we can go through the end of the restriction because only the start would mark a
+      // potential stopping point (maybe a path followed the restriction). We also have to stop
+      // at the start of a simple restriction because it could have been on our path
+
       // if this opposing edge is traversable we enqueue its begin node
-      if (edge_filter(opp_edge) > 0)
+      if (edge_filter(opp_edge) > 0 && !opp_edge->start_restriction() && !opp_edge->restrictions())
         enqueue(edge.endnode());
     }
   }
   // settled nodes + will be settled nodes - duplicated transitions nodes
   reach.inbound =
-      std::min(static_cast<uint32_t>(queue.size() + done.size() - transitions), max_reach);
+      std::min(static_cast<uint32_t>(queue_.size() + done_.size() - transitions_), max_reach);
+
+  // if we didnt make the limit we should do an exact check for whichever directions we didnt meet
+  auto retry_direction = (direction & kOutbound && reach.outbound < max_reach ? kOutbound : 0) |
+                         (direction & kInbound && reach.inbound < max_reach ? kInbound : 0);
+  if (retry_direction) {
+    auto retry_reach = exact(edge, edge_id, max_reach, reader, costing, retry_direction);
+    reach.outbound = std::max(reach.outbound, retry_reach.outbound);
+    reach.inbound = std::max(reach.inbound, retry_reach.inbound);
+  }
 
   return reach;
 }
 
+directed_reach Reach::exact(const valhalla::baldr::DirectedEdge* edge,
+                            const GraphId edge_id,
+                            uint32_t max_reach,
+                            valhalla::baldr::GraphReader& reader,
+                            const std::shared_ptr<sif::DynamicCost>& costing,
+                            uint8_t direction) {
+  // can we even try to expand?
+  directed_reach reach{};
+  if (costing->GetEdgeFilter()(edge) == 0) {
+    return reach;
+  }
+
+  // fake up the input location
+  const baldr::GraphTile* tile = nullptr;
+  const auto* node = reader.GetEndNode(edge, tile);
+  auto ll = node->latlng(tile->header()->base_ll());
+  locations_.Mutable(0)->mutable_ll()->set_lng(ll.first);
+  locations_.Mutable(0)->mutable_ll()->set_lat(ll.second);
+  locations_.Mutable(0)->mutable_path_edges(0)->set_graph_id(edge_id);
+  locations_.Mutable(0)->mutable_path_edges(0)->mutable_ll()->set_lng(ll.first);
+  locations_.Mutable(0)->mutable_path_edges(0)->mutable_ll()->set_lat(ll.second);
```

Comment:
here we take our previously constructed location and modify it to have the LL and id of the edge we want to do the exact expansion from